src/software/thunderscope/gl/layers/gl_referee_info_layer.py

```diff
@@ -0,0 +1,250 @@
+        if tbots_cpp.contains(
+            self.ball_placement_tolerance_circle,
+            tbots_cpp.createPoint(ball_state.global_position),
+        ):
+            self.placement_tolerance_graphic.set_outline_color(
+                self.BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_ON
+            )
+        else:
+            self.placement_tolerance_graphic.set_outline_color(
+                self.BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_OFF
+            )
+
+    def __update_target_visual(self) -> None:
+        """Update the ball placement target graphic to shrink or expand"""
+        if self.shrink_target:
+            self.placement_target_graphic.set_radius(
+                self.placement_target_graphic.radius - 0.01
+            )
+            self.shrink_target = self.placement_target_graphic.radius >= 0
+        else:
+            self.placement_target_graphic.set_radius(
+                self.placement_target_graphic.radius + 0.01
+            )
+            self.shrink_target = (
+                self.placement_target_graphic.radius
+                >= BALL_PLACEMENT_TOLERANCE_RADIUS_METERS
+            )
+
+        # update the count-down graphics
+        if self.cached_referee_info:
+            time_left = max(
+                int(self.cached_referee_info["currentActionTimeRemaining"]) // 1000000,
+                0,
+            )
+            self.ball_placement_countdown_graphic.setData(text=f"{time_left}s")
+
+    def __display_ball_placement_visuals(self, new_placement_point: Point) -> None:
+        """Display ball placement visuals
+        :param new_placement_point: ball placement point
+        """
+        self.ball_placement_point = new_placement_point
+
+        self.placement_tolerance_graphic.set_position(
+            self.ball_placement_point.x_meters,
+            self.ball_placement_point.y_meters,
+        )
+        self.placement_tolerance_graphic.show()
+
+        self.placement_target_graphic.set_position(
+            self.ball_placement_point.x_meters,
+            self.ball_placement_point.y_meters,
+        )
+        self.placement_target_graphic.set_radius(BALL_PLACEMENT_TOLERANCE_RADIUS_METERS)
+        self.placement_target_graphic.show()
+
+        self.robot_avoid_circle_graphic.set_position(
+            self.ball_placement_point.x_meters,
+            self.ball_placement_point.y_meters,
+        )
+        self.robot_avoid_circle_graphic.show()
+
+        self.ball_placement_countdown_graphic.setData(
+            text=f"{BALL_PLACEMENT_TIME_LIMIT_S}s",
+            pos=[
+                self.ball_placement_point.x_meters
+                + BALL_PLACEMENT_ROBOT_AVOID_RADIUS_METERS / 2,
+                self.ball_placement_point.y_meters
+                + BALL_PLACEMENT_ROBOT_AVOID_RADIUS_METERS
+                + 0.1,
+                0,
+            ],
+        )
+        self.ball_placement_countdown_graphic.show()
+
+        self.placement_start_time = time.time()
```

Comment:
AFAICT: we don't seem to use `placement_start_time` anywhere

Reply:
Good call! Removed